File: CODING/4. Stacks/Infix_to_Postfix_conversion.c
/*
Program to convert a Infix Expression into Postfix Expression.....

Time Complexity : O(N)
Space Complexity : O(N)
*/
#include<stdio.h>
#include<ctype.h>

char stack[100], expr[100], postfix[100], *s = expr, x;
int top = -1;

void push(char x) {

		stack[++top] = x;
}

char pop() {

		if(top == -1) {
			return -1;
		}
		else {
			return stack[top--];
		}
}

int Priority(char x) {

	    if(x == '(') {
	        return 0;
	    }
	    if(x == '+' || x == '-') {
	        return 1;
	    }
	    if(x == '*' || x == '/') {
	        return 2;
	    }
	    return -1;
}

int main() {

		int index = 0;

		printf("\nEnter the expression : ");
		scanf("%s", expr);

		while(*s != '\0') {

				if(isalnum(*s)) {
					postfix[index++] = *s;
				}
				else if(*s == '(') {
					push(*s);
				}
				else if(*s == ')') {

					while((x = pop()) != '(') {
						postfix[index++] = x;
					}
				}
				else {

					while(Priority(stack[top]) >= Priority(*s)) {
						postfix[index++] = pop();
		             	
					}
					push(*s);
				}
				++s;
		}

		while(top != -1) {
	   	     postfix[index++] = pop(); 
	   	}
	    printf("\n POSTFIX :: %s ", postfix);

}

File: CODING/4. Stacks/Infix_to_Postfix_conversion.java
/*
Program to convert a Infix Expression into Postfix Expression.....

Time Complexity : O(N)
Space Complexity : O(N)
*/
import java.util.*;

class GFG {

	// A utility function to return precedence of a given operator
	// Higher returned value means higher precedence
	static int Prec(char ch) {

		switch (ch) {

			case '+':
			case '-':
			    return 1;

			case '*':
			case '/':
			    return 2;

			case '^':
			    return 3;
		}
		return -1;
	}

	private static String infixToPostfix(String exp) {

		String result = new String("");
		Stack<Character> stack = new Stack<>();

		for(int i = 0; i < exp.length(); ++i) {

			char ch = exp.charAt(i);
			// If the scanned character is an operand, add it to output.
			if(Character.isLetterOrDigit(ch)) {
				result += ch;
			}
			// If the scanned character is an '(' push it to the stack
			else if(ch == '(') {
				stack.push(ch);
			}
			//  If the scanned character is an ')'  
			// pop and output from the stack until an '(' is encountered.
			else if(ch == ')') {

				while(!stack.isEmpty() && stack.peek() != '(') {
					result += stack.pop();
				}
				stack.pop();
			} 
			// an operator is encountered
			else {
 				while(!stack.isEmpty() && Prec(ch) <= Prec(stack.peek())) {
					result += stack.pop();
				}
				stack.push(ch);
			}
		}
		// pop all the operators from the stack
		while(!stack.isEmpty()) {

			if(stack.peek() == '(') {
				return "Invalid Expression";
			}
			result += stack.pop();
		}
		return result;
	}

	public static void main(String[] args) {
		
		String exp = "a+b*(c^d-e)^(f+g*h)-i";
		System.out.print(infixToPostfix(exp));
	}
}